IdentityCore/src/cache/token/MSIDCommonCredential.m

```diff
@@ -0,0 +1,217 @@
+    hash = hash * 31 + self.secret.hash;
+    hash = hash * 31 + self.target.hash;
+    hash = hash * 31 + self.realm.hash;
+    hash = hash * 31 + self.environment.hash;
+    hash = hash * 31 + self.expiresOn.hash;
+    hash = hash * 31 + self.cachedAt.hash;
+    hash = hash * 31 + self.familyId.hash;
+    hash = hash * 31 + self.homeAccountId.hash;
+    hash = hash * 31 + self.clientInfo.hash;
+    hash = hash * 31 + self.additionalFields.hash;
+    return hash;
+}
+
+#pragma mark - NSCopying
+
+- (nonnull instancetype)copyWithZone:(NSZone *)zone {
+    MSIDCommonCredential *item = [[self class] allocWithZone:zone];
+    item.clientId = [self.clientId copyWithZone:zone];
+    item.credentialType = self.credentialType;
+    item.secret = [self.secret copyWithZone:zone];
+    item.target = [self.target copyWithZone:zone];
+    item.realm = [self.realm copyWithZone:zone];
+    item.environment = [self.environment copyWithZone:zone];
+    item.expiresOn = [self.expiresOn copyWithZone:zone];
+    item.cachedAt = [self.cachedAt copyWithZone:zone];
+    item.familyId = [self.familyId copyWithZone:zone];
+    item.homeAccountId = [self.homeAccountId copyWithZone:zone];
+    item.clientInfo = [self.clientInfo copyWithZone:zone];
+    item.additionalFields = [self.additionalFields copyWithZone:zone];
+    return item;
+}
+
+#pragma mark - JSON
+
+- (nullable instancetype)initWithJSONDictionary:(NSDictionary *)json error:(NSError **)error {
+    MSID_TRACE;
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    if (!json) {
+        MSID_LOG_WARN(nil, @"Tried to decode a credential cache item from nil json");
+        return nil;
+    }
+
+    _json = json;
+
+    _clientId = json[MSID_CLIENT_ID_CACHE_KEY];
+    _credentialType = [MSIDCredentialTypeHelpers credentialTypeFromString:json[MSID_CREDENTIAL_TYPE_CACHE_KEY]];
+    _secret = json[MSID_TOKEN_CACHE_KEY];
+
+    if (!_secret) {
+        MSID_LOG_WARN(nil, @"No secret present in the credential");
+        return nil;
+    }
+
+    _target = json[MSID_TARGET_CACHE_KEY];
+    _realm = json[MSID_REALM_CACHE_KEY];
+    _environment = json[MSID_ENVIRONMENT_CACHE_KEY];
+
+    _expiresOn = [NSDate msidDateFromTimeStamp:json[MSID_EXPIRES_ON_CACHE_KEY]];
+    _cachedAt = [NSDate msidDateFromTimeStamp:json[MSID_CACHED_AT_CACHE_KEY]];
+    _extendedExpiresOn = [NSDate msidDateFromTimeStamp:json[MSID_EXTENDED_EXPIRES_ON_CACHE_KEY]];
+
+    _familyId = json[MSID_FAMILY_ID_CACHE_KEY];
+    _homeAccountId = json[MSID_HOME_ACCOUNT_ID_CACHE_KEY];
+    _clientInfo = [[MSIDClientInfo alloc] initWithRawClientInfo:json[MSID_CLIENT_INFO_CACHE_KEY] error:nil];
+
+    // Additional Fields
+    NSString *speInfo = json[MSID_SPE_INFO_CACHE_KEY];
+    NSMutableDictionary *additionalFields = [NSMutableDictionary dictionary];
+    additionalFields[MSID_SPE_INFO_CACHE_KEY] = speInfo;
+    if ([additionalFields count]) {
+        _additionalFields = additionalFields;
+    }
+
+#pragma unused(error)
```

Comment:
nit: we usually use "__unused" near parameter name.

Reply:
Will it solve compiler warnings?

Reply:
Are these two things equivalent?

Reply:
yes, they are equivalents,  but "__unused" just look more elegant :)

Reply:
@d-col it makes sense to open a PR in our repo and change all of them

Reply:
Example:
`- (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(__unused NSError **)error`